Raw model fields, full encoding support, roundtrip encoding tests

Infinisil opened the pull request:
*   Raw model fields and full encoding support
    
    This commit extends the model with a "raw field" abstraction, which
    allows parametrizing the fully-decoded model by whether some fields
    contain the raw encoded data or not.
    
    This makes a lot of things nicer:
    - We can now have a function that turns model types without the raw
      fields into one that does, but doesn't do any other encoding.
    - We can generate well-behaved Arbitrary instances for the model without
      the raw fields
    - The above two points combined let us generate test model types, encode
      its raw fields, then verify it (which requires the raw fields)
    
    As part of this, this commit also cleans up and fully implements the
    decoding and encoding of such raw data fields, by creating new
    Crypto.Fido2.Model.Binary.{Encoding,Decoding} modules, which contain
    functions for encoding/decoding all binary fields of the model

*   Small model adjustments
    
    When trying to do roundtrip tests for the model encoding, these updates
    were necessary:
    - Make ccdCrossOrigin a required field. It not being set is the same as
      the value False
    - Comment out arcTransports. It was not used, as the webauthn-json library
      doesn't propagate that field

*   Fully implement Arbitrary instances for Model types
*   Add test for model encoding roundtrips

Infinisil commented:
TODO:
- [ ] Comments
- [ ] Roundtrip tests for encoding/decoding of the model types

ErinvanderVeen commented on fido/Crypto/Fido2/Model/JavaScript/Encoding.hs:
```diff
@@ -268,11 +242,11 @@ encodePublicKeyCredentialRequestOptions = encode
 -- Encodes the PublicKeyCredential for attestation, this function is mostly used in the tests where we emulate the
 -- of the client.
-encodeCreatedPublicKeyCredential :: M.PublicKeyCredential 'M.Create -> JS.CreatedPublicKeyCredential
+encodeCreatedPublicKeyCredential :: M.PublicKeyCredential 'M.Create 'True -> JS.CreatedPublicKeyCredential
 encodeCreatedPublicKeyCredential = encode
```
It is rather unfortunate that we would have to first enter the `encodeRaw` function before using the function above. Would it be possible to also have this function accept `M.PublicKeyCredential 'M.Create 'True`?

Infinisil replied:
It would be, but then we'd have to involve singletons with `Bool`s, which doesn't yet seem that necessary

ErinvanderVeen commented on fido/Crypto/Fido2/Model/Binary/Encoding.hs:
```diff
@@ -0,0 +1,183 @@
+  where
+    ccdRawData = M.WithRaw $ LBS.toStrict $ toLazyByteString builder
+
+    builder :: Builder
+    builder =
+      stringUtf8 "{\"type\":"
+        <> jsonBuilder typeValue
+        <> stringUtf8 ",\"challenge\":"
+        <> jsonBuilder challengeValue
+        <> stringUtf8 ",\"origin\":"
+        <> jsonBuilder originValue
+        <> stringUtf8 ",\"crossOrigin\":"
+        <> jsonBuilder crossOriginValue
+        <> stringUtf8 "}"
```
We should probably refer to https://www.w3.org/TR/webauthn-2/#clientdatajson-serialization here as well. I can imagine the later reference to this part of the spec can be missed.

ErinvanderVeen commented on fido/Crypto/Fido2/Model/Binary/Encoding.hs:
```diff
@@ -0,0 +1,183 @@
+      }
+  SGet -> \M.NoAttestedCredentialData -> M.NoAttestedCredentialData
+
+encodeAttestedCredentialData :: M.AttestedCredentialData 'M.Create 'True -> Builder
+encodeAttestedCredentialData M.AttestedCredentialData {..} =
+  Binary.execPut (Binary.putByteString $ M.unAAGUID acdAaguid)
+    <> Binary.execPut (Binary.putWord16be credentialLength)
+    <> Binary.execPut (Binary.putByteString $ M.unCredentialId acdCredentialId)
+    <> Binary.execPut (Binary.putByteString $ M.unRaw acdCredentialPublicKeyBytes)
+  where
+    credentialLength :: Word16
+    credentialLength = fromIntegral $ BS.length $ M.unCredentialId acdCredentialId
```
A reference to the spec would suit here.

ErinvanderVeen commented on fido/Crypto/Fido2/Model/Binary/Encoding.hs:
```diff
@@ -0,0 +1,183 @@
+    bytes :: BS.ByteString
+    bytes = LBS.toStrict $ toLazyByteString builder
+
+    flags :: Word8
+    flags =
+      userPresentFlag
+        .|. userVerifiedFlag
+        .|. attestedCredentialDataPresentFlag
+        .|. extensionsPresentFlag
+      where
+        userPresentFlag = if M.adfUserPresent adFlags then Bits.bit 0 else 0
+        userVerifiedFlag = if M.adfUserVerified adFlags then Bits.bit 2 else 0
+        attestedCredentialDataPresentFlag = case sing @t of
+          SCreate -> Bits.bit 6
+          SGet -> 0
+        extensionsPresentFlag = case adExtensions of
+          Just _ -> Bits.bit 7
+          Nothing -> 0
```
A reference to https://www.w3.org/TR/webauthn-2/#flags would be great here

ErinvanderVeen commented on tests/Spec/Types.hs:
```diff
@@ -74,25 +94,128 @@ instance Arbitrary M.Challenge where
+shuffledSubset :: (Ord a, Bounded a, Enum a) => Gen [a]
+shuffledSubset = subset >>= shuffle . Set.toList
+
+subset :: (Ord a, Bounded a, Enum a) => Gen (Set a)
+subset = Set.fromList <$> sublistOf (Set.toList completeSet)
```
I would assume `toList` produces a list with some ordering, meaning that this subset function doesn't produce all valid subsets. Perhaps shuffling the list of the the completeSet before creating a sublist would work? Or perhaps we don't even want all possible subsets, but then the function should be renamed.

Infinisil replied:
`sublistOf` generates a [subsequence](https://en.wikipedia.org/wiki/Subsequence) of the given list, which is what we need here.

Infinisil commented:
I updated the `Binary.{Encoding,Decoding}` modules to only export the functions that are necessary, and document them properly. Also added specs for the implementations as suggested

ErinvanderVeen commented on fido/Crypto/Fido2/Model/Binary/Decoding.hs:
```diff
@@ -0,0 +1,278 @@
+    DecodingErrorUnexpectedAlgorithmIdentifier JS.COSEAlgorithmIdentifier
+  deriving (Show, Exception)
+
+-- | Webauthn contains a mixture of binary formats. For one it's CBOR and
+-- for another it's a custom format. For CBOR we wish to use the
+-- [cborg](https://hackage.haskell.org/package/cborg) library
+-- and for the custom binary format the [binary](https://hackage.haskell.org/package/binary)
+-- library. However these two libraries don't interact nicely with each other.
+-- Because of this we are specifying the decoders as a 'PartialBinaryDecoder DecodingError',
+-- which is just a function that can partially consume a 'LBS.ByteString'.
+-- Using this we can somewhat easily flip between the two libraries while
+-- decoding without too much nastiness.
```
Good explanation!
